src/crossref.js

```diff
@@ -1,18 +1,272 @@
-import Wikicite from './wikicite';
-
+import Wikicite, { debug } from './wikicite';
+import Citation from './citation';
+import Progress from './progress';
 /* global Services */
+/* global Zotero */
 /* global window */
 
 export default class Crossref{
-	static getCitations() {
-        Services.prompt.alert(
+
+    /**
+     * Get source item citations from CrossRef.
+     * @param {SourceItemWrapper[]} sourceItems - One or more source items to get citations for.
+     */
+    static async addCrossrefCitationsToItems(sourceItems){
+
+        // Make sure that at least some of the source items have DOIs
+        const sourceItemsWithDOI = sourceItems.filter((sourceItem) => sourceItem.getPID('DOI'));
+        if (sourceItemsWithDOI.length == 0){
+            Services.prompt.alert(
+                window,
+                Wikicite.getString('wikicite.crossref.get-citations.no-doi-title'),
+                Wikicite.getString('wikicite.crossref.get-citations.no-doi-message')
+            );
+            return;
+        }
+
+        // Get reference information for items from CrossRef
+        const progress = new Progress(
+            'loading',
+            Wikicite.getString('wikicite.crossref.get-citations.loading')
+        );
+
+        let sourceItemReferences;
+        try{
+             sourceItemReferences = await Promise.all(
+                sourceItemsWithDOI.map(async (sourceItem) => await Crossref.getReferences(sourceItem.doi))
+            );
+        }
+        catch (error){
+            progress.updateLine(
+                'error',
+                Wikicite.getString('wikicite.crossref.get-citations.error-getting-references')
+            );
+            debug(error);
+            return;
+        }
```

Comment:
My intention in this case was that `getReferences` would handle all errors, so these Promises should always be fulfilled, either with a list of references or []. An error here would be unexpected, and in this case we should stop and return early?

> This means that the confirmation message later will always say references have been found for 100% of items.

I don't quite get what you mean here? For the items where we didn't find any references or an error occurred getting the list, the entry in `sourceItemReferences` will be `[]`, so it will be filtered out of `itemsToBeUpdated`. Or am I missing something?


Yeah, the rate limiting should be added, but pragmatically it doesn't seem the biggest issue at the moment. We only fire one request per item (which normally would just be 1 or a few). I tested with 130 items at once and didn't get any errors. Like you say, it'd probably make sense to focus on it after everything else is working.

Reply:
> > This means that the confirmation message later will always say references have been found for 100% of items.
> 
> I don't quite get what you mean here? For the items where we didn't find any references or an error occurred getting the list, the entry in `sourceItemReferences` will be `[]`, so it will be filtered out of `itemsToBeUpdated`. Or am I missing something?

No, I made a mistake — I missed that it was a call to `sourceItems.length`, and you're filtering `sourceItems` at the top of the function. Here's some of why:

> My intention in this case was that `getReferences` would handle all errors, so these Promises should always be fulfilled, either with a list of references or []. An error here would be unexpected, and in this case we should stop and return early?

That makes sense, though I don't agree with your design choice. Here's why: you're implementing your own error handling logic, rather than using the logic provided by the language, and I don't see how it adds value (whereas I do see how it adds complexity). There is already a way to return early from the method in an (unanticipated) error case — don't catch the error, here. And rather than returning `[]` for fetch failures, you can `Promise.reject`.

Here's an alternative: `getReferences` (now much clearer — thank you) should return a `Promise` with the (possibly empty) list of references. If the request fails it should `Promise.reject`. In the highlighted snippet, with `Promise.allSettled`, only the fulfilled promises are further processed. Any unanticipated error causes the whole method to fail; this needs handling somewhere (but not in the method itself).

As I've said before, my goal here is to make review as easy as possible for the maintainer so your work can be merged. Every added layer of indirection makes the code more complicated to review (and maintain). It's often worth it, and often not. I think adding an implicit error handling layer on top of the one in the language is not worth it, here.

Reply:
Overall I agree with what you're saying. I just don't quite see how it would fit perfectly here, but please suggest something I might be missing.

* because `getReferences` can innocently return an empty list if there are no references, it seems a bit weird to `reject` in this case? Or do you mean only for when an error occurs getting the references? But then I don't see the benefit - if `getReferences` would `reject`, we'd do the same as if it returned an empty list anyway?
* If I understand you correctly, basically you're saying to remove the try/catch here and let an error naturally bubble up and stop execution? While that would make the code much clearer, I think the progress bar error message to a user is somewhat helpful?

Reply:
> * because `getReferences` can innocently return an empty list if there are no references, it seems a bit weird to `reject` in this case? Or do you mean only for when an error occurs getting the references?

The latter.

> But then I don't see the benefit - if `getReferences` would `reject`, we'd do the same as if it returned an empty list anyway?

The benefit is that it is clear what is happening. It is obvious that there are error cases, and that the code is handling them.

> * If I understand you correctly, basically you're saying to remove the try/catch here and let an error naturally bubble up and stop execution?

That is the correct understanding, and it is just one option.

> While that would make the code much clearer, I think the progress bar error message to a user is somewhat helpful?

OK, this is good. The code needs to be clear. It also ought to present users with helpful error messages. I think the complication comes because the logic for fetching references is interspersed with the logic for showing the UI.

I'll approach this in a separate message.

Reply:
> While that would make the code much clearer, I think the progress bar error message to a user is somewhat helpful?

Incidentally, are you missing a `progress.close()` before the `return`?

Also how does this whole plugin work: can I do anything else with zotero while the items are being fetched?

My less structured thoughts are below the break — I'll have to come back to this.

---

The simple change is to scrap the error message. `Failed to get citations for item(s) from Crossref` is not _that_ helpful to users. Run the whole method in a `try`/`catch`/`finally` block and then `progress.close()` once, in the `finally`.

I've missed something, in my review, that I should have highlighted earlier: handling the data processing and the user interaction together complicates this code.

I fear the right solution might be uniodiomatic for Zotero: to separate the code that deals with the `progress` from the code doing the fetching. I'd normally solve this in javascript with `addEventListener` and `dispatchEvent`. I see examples of the former in this repository, but none of the latter — I presume they work with zotero but I don't know the conventions.

The real challenge I have concerns the need for user interaction. `Services.prompt.confirm` is called synchronously? So execution of the whole thread is just suspended until it returns? That… that is not how I would expect things to work. Is this the whole program thread? Plugin thread? I… don't like this, at all. Plausible that this dip into zotero development goes no further…

I'm going to have to come back to this, another time.

Reply:
I guess it'll be clearer if you are able to build and test the plugin. But in general yeah, everything is running independently of the main UI. So you can start fetching citations from Crossref then go about doing other stuff in Zotero and the progress box will keep you updated with what's going on. That's why I think it's fairly important - the progress bar is your only way as a user to know your command is still running / what it's doing / when it might finish.

I agree it's a bit ugly having all these progress messages in the code, but that's how the rest of the plugin is structured. Would replacing the progress bar updates with `dispatchEvent` make a big practical difference?

As you point out - `Services.prompt.alert` and `Services.prompt.confirm` are blocking - for when we can't proceeed without user interaction. Here both of these occur more or less towards the start of the method, so it shouldn't be too jarring for the user to need to provide some input given they've just clicked a button to do something (rather than this popping up 2 minutes later).